src/cargo/ops/cargo_rustc/links.rs

```diff
@@ -27,17 +28,41 @@ impl<'a> Links<'a> {
         };
         if let Some(prev) = self.links.get(lib) {
             let pkg = unit.pkg.package_id();
-            if prev.name() == pkg.name() && prev.source_id() == pkg.source_id() {
-                bail!("native library `{}` is being linked to by more \
-                       than one version of the same package, but it can \
-                       only be linked once; try updating or pinning your \
-                       dependencies to ensure that this package only shows \
-                       up once\n\n  {}\n  {}", lib, prev, pkg)
-            } else {
-                bail!("native library `{}` is being linked to by more than \
-                       one package, and can only be linked to by one \
-                       package\n\n  {}\n  {}", lib, prev, pkg)
-            }
+
+            let describe_path = |pkgid: &PackageId| -> String {
+                let dep_path = resolve.path_to_top(pkgid);
+                if dep_path.is_empty() {
+                    String::from("(This is the root-package)")
+                } else {
+                    let mut pkg_path_desc = String::from("(Dependency via ");
+                    let mut dep_path_iter = dep_path.into_iter().peekable();
+                    while let Some(dep) = dep_path_iter.next() {
+                        write!(pkg_path_desc, "{}", dep).unwrap();
+                        if dep_path_iter.peek().is_some() {
+                            pkg_path_desc.push_str(" => ");
```

Comment:
Should this point the other way? The directionality here would entice me to interpret this as "the left depends on the right", but I think it's the other way?

We could also expand this a bit more perhaps like:

```
Package `foo v0.4.0` also links to native library `{}`.
    ... which is depended on by `bar v0.4.0`
    ... which is depended on by `test v0.3.0`
```